### ISSUE TITLE
CoilCoolingWater:addToNode resets ControllerWaterCoil Control Variable

### ISSUE THREAD

**Issue by eringold** [NREL/OpenStudio#4445]
Issue overview
--------------
I want to set the Control Variable of a ControllerWaterCoil prior to assigning the CoilCoolingWater to an AirLoopHVAC. But the value gets reset when the coil is added to the air loop node.

## Current Behavior
```
require 'openstudio'
include OpenStudio::Model

m = Model.new
pl = PlantLoop.new(m)
a = AirLoopHVAC.new(m)
coil = CoilCoolingWater.new(m)

pl.addDemandBranchForComponent(coil)
ctrl = coil.controllerWaterCoil.get
ctrl.setControlVariable("TemperatureAndHumidityRatio")
puts ctrl.controlVariable
#=> TemperatureAndHumidityRatio

coil.addToNode(a.supplyOutletNode)
puts coil.controllerWaterCoil.get.controlVariable
#=> Temperature (this is the default)
```

## Expected Behavior
Keep my assignments! 

## Possible Solution
CoilCoolingWater_Impl::addToNode currently removes and replaces the existing controller: 

https://github.com/NREL/OpenStudio/blob/c4a5d4f5e3d5239a93029131ec61438cfb762b4d/src/model/CoilCoolingWater.cpp#L320-L332

Can it retain any prior assignments to the controller?

 - Version of OpenStudio: 3.2.0+e11f0a08b2

**Comment by eringold** [comment]
OK I guess this isn't a huge deal, since probably in most cases the coil controller inputs aren't exposed until after the coil is added to the loop. And I can somewhat easily re-configure my code to do the same. But I still find a silent override like this kind of jarring.

Besides, why is `oldController` even being removed?

**Comment by jmarrec** [comment]
There is some weirdness with this API but it's been like this since the beginning so we shouldn't change most of it.

* To avoid orphaning controllers, when you remove the coil from the plantloop the controller is removed.
* When you add to a loop, it is trying to be smart for you...
* The ControllerWaterCoil::setWaterCoil method isn't exposed.
* The ControllerWaterCoil Ctor should probably take a WaterToAirComponent as an argument. Right now you can still create one manually, but it's completely useless because you can't assign it to a coil...

The basic idea I think was to completely delegate the creation/removal of the ControllerWaterCoil to other classes, as in you were not supposed to mess with it directly (and it's unfortunate you *somewhat* can)

I think to that effect it's a shame that there aren't a WaterToAirComponent::(control|setControl)Variable method that forwards to the ControllerWaterCoil. Then ControllerWaterCoil could have been omitted completely from the public API. Anyways!


Problem 1:
* At the very least, it should only remove&add the ControllerWaterCoil is the CoilCoolingWater_Impl::addToNode is called with a node on a **PlantLoop** instead of doing it even if you connect to an AirLoop node!
* I don't see why we couldn't just reuse the controller if it already had one instead of removing/creating a new one. @kbenne might though

**Comment by kbenne** [comment]
> At the very least, it should only remove&add the ControllerWaterCoil is the CoilCoolingWater_Impl::addToNode is called with a node on a PlantLoop instead of doing it even if you connect to an AirLoop node!

Well sooner or later the water coil is going to get assigned to both a plant and an air loop, so I guess I don't see the problem with adding the controller on add to either. Although, changing it to only plant doesn't seem like it would cause any harm to me, and it would probably make the logic more simple.

> I don't see why we couldn't just reuse the controller if it already had one instead of removing/creating a new one. @kbenne might though

I agree with that!

> The ControllerWaterCoil Ctor should probably take a WaterToAirComponent as an argument. Right now you can still create one manually, but it's completely useless because you can't assign it to a coil...

I agree the controller should take a WaterToAirComponent on construction. Maybe you have actually stumbled on a place where we can change the API with impunity, since there isn't a valid use case for directly creating a orphan controller.

**Comment by eringold** [comment]
>I agree with that! And if we make that change, wouldn't it address the issue that @eringold is having?

Yes! I'm fine with having the controller created when the coil is assigned to the plant loop, but once that happens (and I can get it and change it's parameters) it should keep those parameters, IMO.

**Comment by jmarrec** [comment]
Some other weird stuff in this API: Lots of methods are returning a boost::optional when there is an IDD default, and don't have a isXXXdefaulted either

https://github.com/NREL/OpenStudio/blob/92796372b7e719241b2912eacbbc8a9907f82a05/src/model/ControllerWaterCoil.hpp#L76

**Comment by jmarrec** [comment]
Haha, I started breaking the API to make the Ctor take an HVACComponent. Only to realize that this can't happen because of `CoilSystemCoolingWaterHeatExchangerAssisted` where we don't populate the water coil on purpose but instead tweak the actuator node in model and add the sensor node on FT.
